src/vs/editor/common/modes.ts

```diff
@@ -392,7 +392,7 @@ export interface ReferenceProvider {
 	/**
 	 * Provide a set of project-wide references for the given position and document.
 	 */
-	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, token: CancellationToken): Location[] | Thenable<Location[]>;
+	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, progress: (value: Location) => void, token: CancellationToken): Location[] | Thenable<Location[]>;
```

Comment:
> So more overhead only if you have to send it over IPC (it probably has to?).

Yes, this has to be serialized to get from the extension host to the main thread.

> if the consumer happens to not use React, he has no way to find out what exactly changed

Based on my work here, not having access to the diff is not a problem. I just instantiate a new ReferencesModel and rerender the UI.

 > TypeScript actually has a concept of Partials. It is a mapped type from an interface to an interface with all properties optional.

Interesting, I did not know that. Good to know!

> I think what you are suggesting is having a class that represents something like an addition to an existing location, identified by an identifier.

More like PartialLocation is a strict subset of Location's fields. In any case, I am not convinced we need a separate type. The "partial" result could just use the `Location` type and leave unknown fields blank (minimum required is an ID). When it has complete information, it just resends that location with the updated location.

```
// Notify progress of partial results
progress(new Location("1", ""));
progress(new Location("2", ""));
progress(new Location("3", ""));

// Later notify progress of full results
progress(new Location("1", "some other data 1"));
progress(new Location("2", "some other data 2"));
progress(new Location("3", "some other data 3"));
```

I don't think we can hash out if JSON patch is the right thing to do here. It is something we will need to figure out but it will require feedback from vscode team. I don't think this issue blocks this PR though. We can submit this using just `Location` and then if we agree to use JSON patch, we can follow up with another change.